exercise4/src/camera_demo/src/actuation_node.cpp

```diff
@@ -0,0 +1,86 @@
+// Listen to the /stop topic and stop the robot when a message is received
+
+#include <chrono>
+#include <memory>
+#include <string>
+#include <vector>
+#include <cactus_rt/tracing.h>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/empty.hpp"
+
+using namespace std::chrono_literals;
+
+using FileSink = cactus_rt::tracing::FileSink;
+using TraceAggregator = cactus_rt::tracing::TraceAggregator;
+using ThreadTracer = cactus_rt::tracing::ThreadTracer;
+
+
+class ActuationNode : public rclcpp::Node
+{
+public:
+  ActuationNode()
+  : Node("actuation_node")
+  {
+    StartTracing("actuation_node", "actuation_node.json");
+    trace_aggregator = nullptr;
+    main_thread_tracer = std::make_shared<ThreadTracer>("main_thread");
+
+    // Create a subscription on the "stop" topic
+    subscription_ = this->create_subscription<std_msgs::msg::Empty>(
+      "stop", 10, std::bind(&ActuationNode::stop_robot, this, std::placeholders::_1));
+  }
+
+private:
+  void StartTracing(const char* app_name, const char* filename) {
+    // Enable the tracing.
+    cactus_rt::tracing::EnableTracing();
+
+    // Create the trace aggregator that will pop the queues and write the events to sinks.
+    trace_aggregator = std::make_unique<TraceAggregator>(app_name);
```

Comment:
You shouldn't be calling this multiple times. There should only be a single instance of this globally otherwise there could be some issues. Also the thread tracers needs to be registered with the trace aggregator otherwise the data won't be popped.

In the interest of time would it be better if we merged this and I can fix the tracing?

Reply:
I forgot to refactor exercise4. Is my usage of the trace aggregators and thread tracers correct in exercise 3?